File: java/RemoveNthNodeFromEndOfList.java
// 19. Remove Nth Node From End of List
// 11ms

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode p = head, q = head;
        int len = 0, idx;
        
        while(p != null) {
            len++;
            p = p.next;
        }
        
        idx = (len - n) + 1;
        len = 0;
        p = head;
        
        while(p != null) {
            len++;
            if(len == idx)
                break;
            
            q = p;
            p = p.next;
        }
        
        if(p == head)
            return head.next;
        else if(p == null)
            q.next = null;
        else
            q.next = p.next;
        
        return head;
    }
}

File: src/problems/019. Remove Nth Node From End of List/RemoveNthNodeFromEndOfList.py
# 19. Remove Nth Node From End of List
# Accepted 56ms

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        
        count = 0
        p = head
        while p != None:
            count += 1
            p = p.next
        n = count - n + 1
        
        count = 1
        p, q = head, dummy
        while p != None:
            if count == n:
                q.next = p.next
                break
            count += 1
            q = p
            p = p.next
        
        return dummy.next
